packages/multichain/src/handlers/wallet-getSession.ts

```diff
@@ -0,0 +1,63 @@
+import type { Caveat } from '@metamask/permission-controller';
+import type { JsonRpcRequest, JsonRpcSuccess } from '@metamask/utils';
+import type { NormalizedScopesObject } from 'src/scope/types';
+
+import { getSessionScopes } from '../adapters/caip-permission-adapter-session-scopes';
+import type { Caip25CaveatValue } from '../caip25Permission';
+import {
+  Caip25CaveatType,
+  Caip25EndowmentPermissionName,
+} from '../caip25Permission';
+
+/**
+ * Handler for the `wallet_getSession` RPC method.
+ *
+ * @param request - The request object.
+ * @param response - The response object.
+ * @param _next - The next middleware function.
+ * @param end - The end function.
+ * @param hooks - The hooks object.
+ * @param hooks.getCaveat - Function to retrieve a caveat.
+ */
+async function walletGetSessionHandler(
+  request: JsonRpcRequest & { origin: string },
+  response: JsonRpcSuccess<{ sessionScopes: NormalizedScopesObject }>,
+  _next: () => void,
+  end: () => void,
+  hooks: {
+    getCaveat: (
+      origin: string,
+      endowmentPermissionName: string,
+      caveatType: string,
+    ) => Caveat<typeof Caip25CaveatType, Caip25CaveatValue>;
+  },
+) {
+  let caveat;
+  try {
+    caveat = hooks.getCaveat(
+      request.origin,
+      Caip25EndowmentPermissionName,
+      Caip25CaveatType,
+    );
+  } catch (e) {
+    // noop
+  }
+
+  if (!caveat) {
+    response.result = { sessionScopes: {} };
+    return end();
```

Comment:
I didn't really find consistent usage throughout the handler implementations I've seen. Not a dumb question at all! The middleware pipeline is flexible, and maybe it shouldn't be so flexible. Do you have a preferred approach/pattern?

Reply:
I think I assumed that handlers would always be used inside of async middleware, because if they were, we would be able to take advantage of the fact that async middleware don't require calling `end`:

```
 * Async middleware have no "end" function. Instead, they "end" if they return
 * without calling "next". Rather than passing in explicit return handlers,
 * async middleware can simply await "next", and perform operations on the
 * response object when execution resumes.
```

Looking back at this, though, I see that 1) handlers aren't middleware themselves and 2) handlers can't assume that they are used inside of async middleware. Maybe when we refactor `json-rpc-engine` to use async middleware everywhere we can simplify this, but it seems like we have to take `end` and call it manually.